src/cfg/File.cc

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "cfg/File.h"
+#include "parser/Tokenizer.h"
+#include "sbuf/Stream.h"
+
+#if HAVE_SYS_STAT_H
+#include <sys/stat.h>
+#endif
+
+Cfg::File::~File()
+{
+    if (fd) {
+        if (isPipe)
+            pclose(fd);
+        else
+            fclose(fd);
+    }
+}
+
+void
+Cfg::File::tryLoadFile()
+{
+    if (filePath[0] == '!' || filePath[0] == '|') {
+        isPipe = true;
+        filePath.erase(0,1);
+    }
+
+    struct stat data;
+    memset(&data, 0, sizeof(data));
+
+    if (::stat(filePath.c_str(), &data) != 0) {
+        int xerrno = errno;
+        throw TextException(ToSBuf("configuration file ", filePath, " error: ", xstrerr(xerrno)), lineInfo());
+    }
+
+    if (S_ISFIFO(data.st_mode))
+        isPipe = true;
```

Comment:
> Why do we want to overwrite what the admin told us?

We do not override.  isPipe defaults to false. Admin can set isPipe=true by configuring a pipe path in squid.conf, OR, we can auto-detect a pipe existence when given a generic path.

Because isPipe is boolean I dropped " !isPipe &&" from this simple if-condition. The outcome does not change. If you insist it can go back.


Background:
  * POSIX requires the use of popen() to access paths prefixed with "|", but leaves handling unspecified in the absence.
  * Squid currently obeys the POSIX requirement at some file loading times, violates it at others. This PR makes all config file loading compliant.

  * Popular software which supports pipes typically performs this stat() check to decide on popen() when given a generic path.
 * Squid currently does not.

 * Several admin are using pipes without explicitly configuring the "|" prefix on paths. They are lucky that their OS fopen(2) supports it.
 * I added this auto-detect to minimize future problems and complaints when that usage moves to OS that are more picky. It will have no effect on working installations, and fix problems some might simply not complained about yet.

Reply:
Official Squid uses popen() for files configured using `|` or `!` prefix and fopen() for all other files. PR behavior is different.

Please interpret my change request and followup comments under the assumption that when an admin says "squid.conf" (no prefix), the admin means "an ordinary file named squid.conf". If you disagree with that assumption, we can discuss it _first_ (ignore everything else then), but it feels wrong to discuss that in the context of a SourceLayout PR that is not supposed to change how Squid works (to the extent possible), especially in controversial areas! I recommend preserving the current behavior instead.

For now, in hope to make progress, I will continue with my assumption in mind.

----

> Why do we want to overwrite what the admin told us?

> We do not override.

An admin configures the file to be a regular file. Based on that configuration, isPipe value is false. This PR code makes isPipe true (in some cases). I called that overwriting the admin decision. You can call it X (i.e. whatever you want), but this PR should not do X.

> Because isPipe is boolean I dropped " !isPipe &&" from this simple if-condition. The outcome does not change.

The correct condition here, especially for a SourceLayout PR, is a `false` condition. In other words, this `if` statement itself should be removed.

Reply:
"foo is foo" removal done in be4e9bd

Filtering of the filePath contents moved to the constructor in 4952a5a. I hope that makes your definition of "proper implementation".